In [4]:
import pandas as pd
READ = "r"
TRAINING_FILE_PATH = '/home/karun/SML/train.txt'
TESTING_FILE_PATH = '/home/karun/SML/test-public.txt'

In [2]:

# Mount google drive and access the uploaded train and test data

from google.colab import drive
drive.mount("/content/drive")

TRAINING_FILE_PATH = '/content/drive/My Drive/train.txt'
TESTING_FILE_PATH = '/content/drive/My Drive/test-public.txt'

TRAINING_FILE_PATH = '/content/drive/My Drive/train.txt'

ModuleNotFoundError: No module named 'google.colab'

In [27]:
train_data = open(TRAINING_FILE_PATH, READ)

**Read the data in train file and conduct simple analysis**

In [49]:


record_count = 0

# stores the mapping from a person to the people they follow
follows = {}

unique_ids = set({})
total_follows_count = 0
min = 100000000
max = -1

# tracks the number of people who follow nobody
zero_follows_records = 0

for entry in train_data:

  data = entry.strip().split('\t')
  # convert string into an array of ids
  ids = [int(id_string) for id_string in data]

  # add all the entries to the SET to find the unique ids in the training data
  unique_ids.update(ids)

  # extract the user in question
  user_id = ids[0]

  # extract the people who the user follows, by removing the user-id from the ids
  ids = ids[1:]

  # get the count of people who the user follows
  follows_count = len(ids)

  # update total records count
  record_count = record_count + 1

  # maps user to the people they follow
  follows[user_id] = ids
  total_follows_count = total_follows_count + follows_count

  # if the people a user follows is zero, update the zero_follow record count
  if follows_count == 0:
    zero_follows_records += 1

  # calculate min and max followed count
  if follows_count > max:
    max = follows_count
  if follows_count < min:
    min = follows_count


print("Train data statistics")
print('Total records         : ', record_count)
print('Total unique ids      : ', len(unique_ids))
print('Total linkages        : ', total_follows_count)
print('Max no. of people followed by a single person  : ', max)
print('Min no. of people followed by a single person  : ', min)
print('No. of users who follow nobody : ', zero_follows_records)



Train data statistics
Total records         :  20000
Total unique ids      :  4867136
Total linkages        :  24004361
Max no. of people followed by a single person  :  764195
Min no. of people followed by a single person  :  0
No. of users who follow nobody :  430


There are 20000 records <br/>
There are 4,867,136 unique ids <br/>
There are a total of 24,004,361 edges in the graph<br/>
There are 430 records with no linkages<br/>
The maximum degree of a node is 764195<br/>

In [7]:
# Create a reverse mapping from a user to the people who follow them

# stores the mapping from a person to the people who follows them
followers = {}


# random variable(no of people follow = input)
follows_count_frequency = {}

for follower in follows:

  # get the number of people a person follows
  follows_count = len(follows[follower])

  # calculate a distribution of no of followed persons for different people
  # KEY - #followed people    VALUE - #followers who have KEY(#followed) number of followed people
  if follows_count in follows_count_frequency:
    follows_count_frequency[follows_count] = follows_count_frequency[follows_count] + 1
  else:
    follows_count_frequency[follows_count] = 1


  # get all the people a follower, follows
  followed_persons = follows[follower]
  # create a reverse mapping
  for followed_person in followed_persons:
    if followed_person in followers:
      followers[followed_person].append(follower)
    else:
      followers[followed_person] = [follower]



print(follows_count_frequency)

{143: 28, 21: 79, 764195: 1, 297: 14, 3808: 1, 4519: 1, 333: 17, 227: 24, 258: 18, 1225: 6, 87: 55, 8414: 1, 314: 9, 1148: 2, 0: 430, 1935: 4, 11757: 1, 369: 10, 6230: 1, 226: 24, 366: 10, 1612: 1, 101: 41, 698: 4, 113: 50, 10357: 1, 1458: 1, 7971: 1, 389: 9, 1: 247, 25: 100, 133: 39, 107: 57, 222: 22, 120: 38, 58: 68, 430: 10, 148: 35, 5849: 1, 363: 12, 197: 32, 52: 79, 206: 25, 89: 64, 138: 49, 33: 82, 1274: 2, 57: 78, 14576: 1, 124: 38, 67: 64, 357: 11, 97: 53, 213: 26, 185: 23, 737: 5, 78: 73, 1526: 2, 175: 35, 73: 67, 122: 60, 131: 32, 316: 10, 2455: 3, 220: 23, 26: 84, 117: 40, 21104: 1, 88: 61, 1699: 1, 902: 4, 40: 69, 2972: 2, 188: 18, 56: 59, 92: 60, 55: 64, 47: 73, 118: 40, 1012: 2, 7579: 1, 1571: 1, 4229: 3, 31: 93, 767: 1, 6562: 1, 554: 4, 2891: 1, 969: 1, 3795: 1, 11: 86, 64: 76, 162: 38, 116: 41, 2002: 5, 770: 5, 68772: 1, 38: 95, 155: 42, 32035: 1, 27: 82, 76: 57, 24: 91, 114: 39, 43: 82, 183: 23, 3: 120, 6: 110, 61: 52, 34: 86, 15: 80, 3377: 1, 48: 71, 119: 35, 296: 15,

In [13]:
from collections import OrderedDict

print(len(follows_count_frequency))
sorted_follows_count = OrderedDict(sorted(follows_count_frequency.items(), key=lambda t: t[0]))
print(sorted_follows_count)



3077
OrderedDict([(0, 430), (1, 247), (2, 138), (3, 120), (4, 108), (5, 96), (6, 110), (7, 81), (8, 80), (9, 84), (10, 65), (11, 86), (12, 83), (13, 69), (14, 108), (15, 80), (16, 90), (17, 89), (18, 102), (19, 86), (20, 77), (21, 79), (22, 86), (23, 101), (24, 91), (25, 100), (26, 84), (27, 82), (28, 74), (29, 78), (30, 82), (31, 93), (32, 81), (33, 82), (34, 86), (35, 76), (36, 82), (37, 65), (38, 95), (39, 78), (40, 69), (41, 76), (42, 85), (43, 82), (44, 66), (45, 61), (46, 72), (47, 73), (48, 71), (49, 74), (50, 76), (51, 74), (52, 79), (53, 63), (54, 79), (55, 64), (56, 59), (57, 78), (58, 68), (59, 73), (60, 66), (61, 52), (62, 72), (63, 58), (64, 76), (65, 56), (66, 54), (67, 64), (68, 78), (69, 78), (70, 59), (71, 58), (72, 67), (73, 67), (74, 52), (75, 53), (76, 57), (77, 56), (78, 73), (79, 59), (80, 69), (81, 59), (82, 52), (83, 54), (84, 48), (85, 47), (86, 60), (87, 55), (88, 61), (89, 64), (90, 62), (91, 55), (92, 60), (93, 52), (94, 57), (95, 63), (96, 48), (97, 53), (9

In [8]:
!pip install neo4j

     |████████████████████████████████| 67 kB 1.6 MB/s eta 0:00:01
  Created wheel for neo4j: filename=neo4j-4.1.1-py3-none-any.whl size=94668 sha256=22edcc65df9e8e4b6655401057bf800928bda5341784148f5d4698ebb0b212f7
  Stored in directory: /home/karun/.cache/pip/wheels/d3/00/f4/fc8b193e3a983822202a43c13d6c4d4e71fbd6d14fca0937f1
Successfully built neo4j
You should consider upgrading via the '/home/karun/PycharmProjects/IntentClassification/venv/bin/python -m pip install --upgrade pip' command.


In [ ]:
import logging
from neo4j import GraphDatabase
from neo4j.exceptions import ServiceUnavailable

class App:

    def __init__(self, uri, user, password):
        self.driver = GraphDatabase.driver(uri, auth=(user, password))

    def close(self):
        # Don't forget to close the driver connection when you are finished with it
        self.driver.close()

    def create_follow_relation(self, person1_name, person2_name):
        with self.driver.session() as session:
            # Write transactions allow the driver to handle retries and transient errors
            result = session.write_transaction(
                self._create_and_return_follow_relation, person1_name, person2_name)

    @staticmethod
    def _create_and_return_follow_relation(tx, person1_name, person2_name):

        # To learn more about the Cypher syntax,
        # see https://neo4j.com/docs/cypher-manual/current/

        # The Reference Card is also a good resource for keywords,
        # see https://neo4j.com/docs/cypher-refcard/current/

        query = (
            "MERGE (p1:Person { name: $person1_name }) "
            "MERGE (p2:Person { name: $person2_name }) "
            "MERGE (p1)-[:FOLLOWS]->(p2) "
            "RETURN p1, p2"
        )
        result = tx.run(query, person1_name=person1_name, person2_name=person2_name)
        try:
            return [{"p1": record["p1"]["name"], "p2": record["p2"]["name"]}
                    for record in result]
        # Capture any errors along with the query and data for traceability
        except ServiceUnavailable as exception:
            logging.error("{query} raised an error: \n {exception}".format(
                query=query, exception=exception))
            raise

    def find_person(self, person_name):
        with self.driver.session() as session:
            result = session.read_transaction(self._find_and_return_person, person_name)
            for record in result:
                print("Found person: {record}".format(record=record))
                return record

    @staticmethod
    def _find_and_return_person(tx, person_name):
        query = (
            "MATCH (p:Person) "
            "WHERE p.name = $person_name "
            "RETURN p.name AS name"
        )
        result = tx.run(query, person_name=person_name)
        return [record["name"] for record in result]


# See https://neo4j.com/developer/aura-connect-driver/ for Aura specific connection URL.
# scheme = "neo4j"  # Connecting to Aura, use the "neo4j+s" URI scheme
# host_name = "example.com"
# port = 7687
# url = "{scheme}://{host_name}:{port}".format(scheme=scheme, host_name=host_name, port=port)
# user = "<Username for Neo4j database>"
# password = "<Password for Neo4j database>"
uri = "neo4j://localhost:7687"
driver = GraphDatabase.driver(uri, auth=("neo4j", "karun"))


app = App(uri, "neo4j", "karun")
app.create_follow_relation("Alice", "David")
app.create_follow_relation("Bill", "Alice")
a = app.find_person("Alice")


for follower in follows:
    # get all the people a follower, follows
    followed_persons = follows[follower]
    # create a follows relation
    for followed_person in followed_persons:
        app.create_follow_relation(follower, followed_person)

app.close()

Found person: Alice


In [48]:
from neo4j import GraphDatabase

uri = "neo4j://localhost:7687"
driver = GraphDatabase.driver(uri, auth=("neo4j", "karun"))

def get_followed_persons_of(tx, name):
    friends = []
    result = tx.run("MATCH (a:Person)-[:FOLLOWS]->(f) "
                         "WHERE a.name = $name "
                         "RETURN f.name AS friend", name=name)
    for record in result:
        friends.append(record["friend"])
    return friends

with driver.session() as session:
    friends = session.read_transaction(get_followed_persons_of, "Bill")
    for friend in friends:
        print(friend)

driver.close()

Alice


In [38]:
# Print out some statistics of the followed person and their followers

print('Total count of people who have followers ', len(followers))

min = 100000000
max = -1

all_follower_counts = []

# random variable(no of followers = input)
follower_count_frequency = {}

for followed_person in followers:

  # get the number of followers of a person
  followers_count = len(followers[followed_person])

  all_follower_counts.append(followers_count)

  # calculate a distribution of no of followers of different followed people
  # KEY - #followers    VALUE - #followed people who have KEY(#followers) number of followers
  if followers_count in follower_count_frequency:
    follower_count_frequency[followers_count] = follower_count_frequency[followers_count] + 1
  else:
    follower_count_frequency[followers_count] = 1
    
  # calculate min and max count of a persons followers
  if followers_count > max:
    max = followers_count
  if follows_count < min:
    min = followers_count

print('The max no. of followers that a person is having', max)
print('The min no. of followers that a person is having', min)

# df = pd.DataFrame(all_follows_count, columns=['follows'])
# pd.value_counts(df['follows']).plot.bar()

print(follower_count_frequency)
sorted_followers_count = dict(sorted(follower_count_frequency.items(), key=lambda item: item[1], reverse=True))
print(sorted_followers_count)

Total count of people who have followers  4867136
The max no. of followers that a person is having 4841
The min no. of followers that a person is having 37
{37: 3621, 76: 957, 28: 5563, 107: 427, 57: 1728, 62: 1457, 44: 2825, 330: 14, 95: 532, 25: 6608, 79: 862, 172: 111, 305: 19, 708: 1, 313: 23, 50: 2220, 54: 1990, 1605: 1, 538: 4, 82: 797, 533: 7, 657: 1, 659: 3, 568: 6, 463: 9, 418: 12, 170: 114, 272: 24, 111: 378, 269: 37, 903: 3, 613: 4, 215: 52, 186: 75, 209: 62, 163: 150, 188: 79, 34: 4315, 755: 4, 895: 1, 1283: 1, 424: 8, 45: 2822, 184: 85, 6: 74822, 333: 19, 301: 23, 1117: 3, 401: 10, 118: 322, 736: 3, 112: 363, 47: 2433, 153: 159, 208: 61, 283: 30, 365: 14, 407: 8, 545: 7, 768: 3, 152: 134, 468: 9, 68: 1175, 1068: 1, 1337: 2, 777: 2, 346: 8, 527: 2, 246: 47, 196: 54, 257: 40, 432: 10, 46: 2540, 766: 2, 65: 1358, 409: 6, 274: 37, 264: 23, 33: 4504, 49: 2224, 85: 741, 91: 597, 166: 117, 58: 1628, 22: 8095, 64: 1314, 904: 2, 27: 5940, 39: 3320, 20: 9428, 425: 8, 109: 364, 420: 

In [ ]:

test_data = open(TESTING_FILE_PATH, READ)

next(test_data)

for entry in test_data:
  entry = entry.strip()
  test_points = entry.split('\t')
  print(test_points)
  print(follows[int(test_points[1])])
  print(follows[int(test_points[2])])
  print(followers[int(test_points[1])])
  # print(followers[int(test_points[2])])
  break


['1', '3563811', '3600160']
[4008078, 3306919, 180302]
[1436216, 1339664, 4088731, 3844858, 3871554, 1311817, 2981537, 3438576, 3509903, 3914135, 1927211, 1912031, 2961552, 4408813, 2845215, 1115799, 3597029, 1847461, 3263148, 2538150, 2790006, 1063857, 2498087, 4801471, 3002746, 1086912, 679540, 3038385, 3451194]
[787039, 3519329, 2472993, 3274877, 1541304, 673835, 1405689, 2401069, 1628939, 4016137, 4008078, 3017264, 2331231, 127885, 3355752, 1523982, 2166423, 1120944, 3638703, 2558493, 4027460]


In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# matplotlib histogram
plt.hist(df['follows'], color = 'blue', edgecolor = 'black',
         bins = int(180/5))

# seaborn histogram
sns.distplot(df['follows'], hist=True, kde=False, 
             bins=int(180/5), color = 'blue',
             hist_kws={'edgecolor':'black'})
# Add labels
plt.title('')
plt.xlabel('')
plt.ylabel('')